Added commit / rollback hooks

jackd opened the pull request:
As requests in #265, I couldn't get any web tests working so I wouldn't at all be surprised if they're broken... but might be best to seek feedback on the overall design before getting too finicky there.

simolus3 commented on sqlite3/assets/wasm/bridge.h:
```diff
@@ -50,10 +50,12 @@ import_dart("function_xInverse") extern void dartXInverse(
-                                                        const char *db,
-                                                        const char *table,
-                                                        sqlite3_int64 rowid);
+import_dart("function_update_hook") extern void dartUpdateHook(void *id, int kind,
```
Please keep the existing name unchanged, newer versions of `package:sqlite3` need to be compatible with older WebAssembly bundles too.

simolus3 commented on sqlite3/lib/src/implementation/database.dart:
```diff
@@ -228,6 +231,8 @@ base class DatabaseImplementation implements CommonDatabase {
       listener.close();
     }
```
I think we should also close the rollback listeners here.

simolus3 commented on sqlite3_web/lib/src/database.dart:
```diff
@@ -50,6 +50,12 @@ abstract class Database {
+  ///
+  /// Updates are only sent across worker channels while a subscription to this
+  /// stream is active.
+  Stream<void> get rollbacks;
```
It's not clear how we would best translate this feature to `sqlite3_web`. We can't provide commit hooks that would be able to revert a transaction because that would require an async roundtrip between the worker and the main page.

We could provide a "read-only" `Stream<void> get commits` stream that installs a `() => true` filter reporting events if necessary. But I'm also fine with not exposing this in `sqlite3_web` until it is needed (`sqlite3_web` also provides the option of compiling custom workers where users have full control over the database, so they could install commit / rollback filters on the database if they need to).

jackd commented:
Terribly sorry, work got hectic for a bit, then Christmas... and now I'm heading away for a few weeks. Just popping by to say this'll be near the top of the list of things I want to get onto when I'm back :)

jackd commented:
@simolus3 I've reverted the breaking name change and added a `commits` stream that should be compatible with a custom `commitFilter`. I believe the issue I'm having running web tests is down to me being on WSL. I'm not going to pretend to understand what's going on with the wasm code, just tried to make it look similar to updates... but that's obviously not working.

simolus3 commented:
Thanks for following up on this! The web tests will require a new version of the WebAssembly module. We strip functions that we aren't calling in Dart out of the module to reduce its size. So every time that we add new functions possibly called in Dart, we need to preserve backwards compatibility with older modules that don't have those functions.

The setup to compile the modules is quite complex and I have no idea on how it might be done on Windows. I'll compile a module that has those functions, that was we can get the web part ready.

simolus3 reviewed (approved):
LGTM! I've moved some bits around to re-use code between the different types of streams.

Thank you for your contribution!
